Support form submissions in the ViewTransitions router (#8963)
withastro/astro: .changeset/many-weeks-sort.md

---
'astro': minor
---

> Form support in ViewTransitions router

Reviewer: ```suggestion
Form support in View Transitions router
```

> The `<ViewTransitions />` router can now handle form submissions, allowing you to retain stateful UI. This feature is opt-in for semver reasons and can be enabled by adding the `handleForms` prop to the component like so:

Reviewer: ```suggestion
The `<ViewTransitions />` router can now handle form submissions, allowing you to trigger in-page transitions from a form submission while keeping page state. This feature is opt-in and can be enabled by adding the `handleForms` prop to the routing component:
```
I'm gonna need you to sell forms to me harder than this, Matthew.  😅 

Can you do better than "retain stateful UI"? Something that more walks through the (exciting!) process itself? People are always asking about forms, can we hype this in a more "give the people what they want" kind of way?

Author: All right, I needed the pep-talk. Pushed with a bit more, trying to emphasize that this addition completes the circle of everything you need to stay in the client if that's what you want.

```astro
---
// src/layouts/MainLayout.astro
import { ViewTransitions } from 'astro:transitions';
---

<html>
  <head>
    <!-- ... -->
    <ViewTransitions handleForms />
  </head>
  <body>
    <!-- ... -->
  </body>
</html>
```

> Like with links, you can opt-out of the router handling a submission on a per-form basis with the `data-astro-reload` property:

Reviewer: ```suggestion
Just as with links, if you don't want the routing handling a form submission, you can opt out on a per-form basis with the `data-astro-reload` property:
```

```astro
---
// src/components/Contact.astro
---

> <form class="contact-form" action="/request" method="post" data-astro-reload>

Reviewer: ```suggestion
<form class="contact-form" action="/request" method="POST" data-astro-reload>
```
Should these now be capitalized?

Author: I think it doesn't matter, but MDN documents them as lower case: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#method

Also vscode completions suggest lower case.

Reviewer: OK, so we changed *our* case for v3 when writing JS in the frontmatter, but because this is just in HTML, we can use the lower case?  That's fine by me!

  <!-- ...-->
</form>
```

> Form support works on post `method="get"` and `method="post"` forms.

Reviewer: ```suggestion
Form support works on post `method="GET"` and `method="POST"` forms.
```

---
'astro': minor
---

Form support in View Transitions router

> The `<ViewTransitions />` router can now handle form submissions, allowing the same animated transitions and stateful UI retention you already get from `<a>` links to occur on form posts. With this added in Astro you can have animations in all of these scenarios:
> 
> - Clicking links between pages.
> - Making stateful changes in forms (ie updating site preferences).
> - Manually triggered navigate via the `navigate()` API.
> 
> This feature is opt-in for semver reasons and can be enabled by adding the `handleForms` prop to the component like so:

Reviewer: ```suggestion
The `<ViewTransitions />` router can now handle form submissions, allowing the same animated transitions and stateful UI retention on form posts that are already available on `<a>` links. With this addition, your Astro project can have animations in all of these scenarios:

- Clicking links between pages.
- Making stateful changes in forms (i.e. updating site preferences).
- Manually triggering navigation via the `navigate()` API.

This feature is opt-in for semver reasons and can be enabled by adding the `handleForms` prop to the `<ViewTransitions /> component:
```
Just suggested some tiny word changes. 

And just want to double check: e.g. should be used if the thing in brackets is one example ( of many possibilities), and i.e. is used when it's literally exactly the thing it is, just rephrased. 

So, if "making stateful changes **=** "updating site preferences", then i.e. is fine. Otherwise, that should be e.g.

Author: Ok, so this is one example, so I'll change to e.g. after merging.
